### ISSUE TITLE
FactorSet should accept any type of Factor

### ISSUE THREAD

**Issue by ankurankan** [pgmpy/pgmpy#733]
Currently `FactorSet` only accepts `DiscreteFactor`. Should accept any subclass of `BaseFactor`.

**Comment by khalibartan** [comment]
@ankurankan How can we create hash values for continuous case factors?

**Comment by ankurankan** [comment]
@khalibartan We can create a string with type of distribution and arguments and then compute it's hash. For example for a gaussian with some mean and cov, we can do something like:
```python
hash('gaussian' + str(mean_values) + str(cov_values))
```

**Comment by raghavg7796** [comment]
@khalibartan @ankurankan I created hash for Gaussian and Canonical but how to create hash in case of continuous Factors with a given pdf. And also how to set factorset_divide in case of gaussian or canonical as it requires identity which raises Singular Error in case of Gaussian and Canonical.

**Comment by ankurankan** [comment]
@raghavg7796 We don't have a standard form for `1 / Normal`. So, to implement that you will need to use the pdf function. But I don't think we will be needing this, so you can throw a `NotImplementedError` if someone calls divide on a factorset having gaussian distributions.

**Comment by raghavg7796** [comment]
@ankurankan How to create hash for user defined pdfs.

**Comment by ankurankan** [comment]
@raghavg7796 The only way I can think of is to generate some values from the pdf and use that along with the variable names to compute the hash. So, like generating probability values for some fixed set of inputs and then converting it to a string and then using the `hash` function. But you will have to make sure to handle the case if the pdf is not defined at some values. Also you will need to consider rounding errors.

And this method might have some hash collisions but I can't think of a better way right now. And the collisions doesn't seem that probable.

**Comment by raghavg7796** [comment]
@ankurankan Can we use UnbiasedDiscretizer for creating hashes??

**Comment by ankurankan** [comment]
@raghavg7796 Yes you should be able to do that. But I am not really sure about the it's status (if it works in case of multivariate distributions). @khalibartan @yashu-seth Opinion ?

**Comment by yashu-seth** [comment]
@raghavg7796 ```UnbiasedDiscretizer``` works only for univariate distributions but user defined pdfs can be multivariate.

**Comment by raghavg7796** [comment]
@yashu-seth @ankurankan So to create hash, we will need to test on some pre-decided input. Is that the only way then ?

**Comment by yashu-seth** [comment]
@raghavg7796 I think we can put a question on stackoverflow and see if we get some ideas from there.

**Comment by raghavg7796** [comment]
@yashu-seth Please can you ask framing the question in a better way.

**Comment by yashu-seth** [comment]
@raghavg7796 Okay. I will put a question and see if we get something.

**Comment by raghavg7796** [comment]
@yashu-seth Did you find a possible solution for this on stackoverflow ?

**Comment by yashu-seth** [comment]
@raghavg7796 No, I could not find anything. . And please hold this issue as there would be certain significant changes with the continuous module.
